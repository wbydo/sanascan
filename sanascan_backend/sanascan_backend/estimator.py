from typing import List, cast, Optional

from .word import Word, TagWord
from .lang_model import LangModel
from .key import Key
from .node import Node, RootNode, EOSNode
from .vocabulary import Vocabulary


class Estimator:
    lang_mode: LangModel
    vocab: Vocabulary
    wait_child: List[List[Node]]  # あとで変数名変えたい
    key: Key
    eos_node: Optional[EOSNode]
    result: Optional[List[Word]]

    def __init__(self, lm: LangModel) -> None:
        self.lang_model = lm
        self.vocab = lm.create_vocabrary()

        self.reset()

    def add(self, key: Key) -> None:
        assert len(key) == 1

        self.wait_child.append([])
        self.key += key

        for word, subkey in self.vocab.get_by_key(self.key):
            candidates_idx = len(self.key) - len(subkey) - 1
            assert candidates_idx >= 0
            candidates = self.wait_child[candidates_idx]
            node = Node(word, candidates, self.lang_model)
            self.wait_child[-1].append(node)

        scores = [node.score for node in self.wait_child[-1]]
        max_node = self.wait_child[-1][scores.index(max(scores))]
        self.result = max_node.sentence[1:]

    def finish(self) -> None:
        self.eos_node = EOSNode(self.wait_child[-1], self.lang_model)
        self.result = self.eos_node.sentence[1:-1]

    def reset(self) -> None:
        self.result = None
        self.wait_child = [[]]
        self.wait_child[0].append(RootNode())

        s_tag = cast(
            TagWord,
            self.wait_child[0][0]._word,
        )

        self.key = Key([s_tag])
