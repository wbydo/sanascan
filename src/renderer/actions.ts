import * as types from "./types";

export interface Action {
  type: string;
}

export const increment: () => Action = () => {
  return {
    type: types.INCREMENT,
  };
};
