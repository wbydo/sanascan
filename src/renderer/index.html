<!DOCTYPE html>
<html lang="ja">
<head>
  <meta charset="utf-8">
  <title>YACOS</title>
  <link rel="stylesheet" href="css/style.css">
</head>
<body>
  <div id="root"></div>
</body>
</html>
