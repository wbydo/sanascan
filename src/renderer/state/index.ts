import { createStore, applyMiddleware } from "redux";
import { combineReducers, Dispatch } from "redux";

import { composeWithDevTools } from "redux-devtools-extension";

import { reducer as configWindowReducer } from "./configWindow";
import { operations as configWindowOperations } from "./configWindow";

import { reducer as cursolReducer } from "./cursol";

import { reducer as estimatorReducer } from "./estimator";
import { middleware as estimatorMiddleware } from "./estimator";
import { operations as estimatorOperations } from "./estimator";

import { reducer as timerReducer } from "./timer";
import { middleware as timerMiddleware } from "./timer";

export interface RootState {
  configWindow: {
    isActive: boolean;
    scanSpeed: number;
  };
  cursol: {
    activeColumn: number;
  };
  timer: {
    id: number | null;
    isActive: boolean;
    scanSpeed: number;
  };
  estimator: {
    id: number | null;
    result: string;
  };
}

const middlewares = [
  timerMiddleware,
  estimatorMiddleware,
];

const enhancer = composeWithDevTools(
  applyMiddleware(...middlewares),
);

export const reducer = combineReducers({
  configWindow: configWindowReducer,
  cursol: cursolReducer,
  estimator: estimatorReducer,
  timer: timerReducer,
});

export const store = createStore(
  reducer,
  enhancer,
);

export const operations = (dispatch: Dispatch) => {
  return {
    configureWindowOpen: configWindowOperations.windowOpen(dispatch),
    resetEstimator: estimatorOperations.reset(dispatch),
    sendKey: estimatorOperations.sendKey(dispatch),
    startFetchEstimatorId: estimatorOperations.fetchId(dispatch),
  };
};
