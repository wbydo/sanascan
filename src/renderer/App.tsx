import * as React from "react";
import {connect} from "react-redux";
import {Dispatch} from "redux";

import CharacterBoard from "./CharacterBoard";
import {Action, asyncIncrement} from "./actions";

interface Props {
  asyncIncrement: () => void;
}

type MapDispatchToProps = (dispatch: Dispatch<Action>) => Props;

class App extends React.Component<Props, {}> {
  public render() {
    return(
      <div>
        <CharacterBoard />
        <button type="button" onClick={this.handleOnClick}>
          async
        </button>
      </div>
    );
  }

  public handleOnClick = (event: React.MouseEvent): void => {
    return this.props.asyncIncrement();
  }
}

const mdp: MapDispatchToProps = (dispatch) => {
  return {
    asyncIncrement: () => dispatch(asyncIncrement()),
  };
};

export default connect(null, mdp)(App);
