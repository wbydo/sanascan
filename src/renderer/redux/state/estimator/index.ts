import { reducer } from "./reducers";
import * as actions from "./actions";

export {
  actions,
  reducer,
};
