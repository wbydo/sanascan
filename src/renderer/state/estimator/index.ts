import * as actions from "./actions";
import middleware from "./middleware";

export const estimatorActions = actions;
export const estimatorMiddleware = middleware;
