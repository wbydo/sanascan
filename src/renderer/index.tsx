import * as React from "react";
import * as ReactDOM from "react-dom";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";

import App from "./App";
import reducer from "./reducers";
import rootSaga from "./sagas";

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  reducer,
  applyMiddleware(sagaMiddleware),
);

sagaMiddleware.run(rootSaga);

ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById("root"),
);
