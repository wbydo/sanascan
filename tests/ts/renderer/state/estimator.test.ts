import snapshotDiff from "snapshot-diff";

import { reducer } from "sanascan/renderer/state";
import { actions } from "sanascan/renderer/state/estimator";

import { initialState } from "./initialState";

test("idが設定できる", () => {
  const id = 1234;
  const result = reducer(
    initialState,
    actions.setId(id),
  );
  expect(
    snapshotDiff( initialState, result),
  ).toMatchSnapshot();
});

test("resultを設定できる", () => {
  const content = "あらゆる現実を全て自分の方へねじ曲げたのだ";
  const result = reducer(
    initialState,
    actions.setResult(content),
  );
  expect(
    snapshotDiff( initialState, result),
  ).toMatchSnapshot();
});
