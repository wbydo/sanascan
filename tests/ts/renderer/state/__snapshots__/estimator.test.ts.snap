// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idが設定できる 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,11 +12,11 @@
    \\"developerMode\\": Object {
      \\"estimatorIsActive\\": false,
      \\"isActive\\": true,
    },
    \\"estimator\\": Object {
-     \\"id\\": null,
+     \\"id\\": 1234,
      \\"result\\": \\"\\",
    },
    \\"timer\\": Object {
      \\"isActive\\": false,
      \\"scanSpeed\\": 500,"
`;

exports[`resultを設定できる 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,11 +13,11 @@
      \\"estimatorIsActive\\": false,
      \\"isActive\\": true,
    },
    \\"estimator\\": Object {
      \\"id\\": null,
-     \\"result\\": \\"\\",
+     \\"result\\": \\"あらゆる現実を全て自分の方へねじ曲げたのだ\\",
    },
    \\"timer\\": Object {
      \\"isActive\\": false,
      \\"scanSpeed\\": 500,
    },"
`;
